flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/MiniClusterITCase.java

```diff
@@ -337,6 +347,262 @@ public void testSchedulingAllAtOnce() throws Exception {
+	 * A sender that does not exit until all receivers are running.
+	 */
+	public static class CountDownLatchedSender extends AbstractInvokable {
+
+		private static CountDownLatch latch;
+
+		static void setLatch(CountDownLatch latch) {
+			CountDownLatchedSender.latch = latch;
+		}
+
+		public CountDownLatchedSender(Environment environment) {
+			super(environment);
+		}
+
+		@Override
+		public void invoke() throws Exception {
+			RecordWriter<IntValue> writer = new RecordWriter<>(getEnvironment().getWriter(0));
+
+			try {
+				writer.emit(new IntValue(42));
+				writer.emit(new IntValue(1337));
+				writer.flushAll();
+			} finally {
+				writer.clearBuffers();
+				latch.await();
+			}
+		}
+	}
+
+	/**
+	 * A receiver that counts down the latch on running.
+	 */
+	public static class CountDownLatchedReceiver extends AbstractInvokable {
+
+		private static CountDownLatch latch;
+
+		static void setLatch(CountDownLatch latch) {
+			CountDownLatchedReceiver.latch = latch;
+		}
+
+		public CountDownLatchedReceiver(Environment environment) {
+			super(environment);
+		}
+
+		@Override
+		public void invoke() throws Exception {
+			latch.countDown();
+
+			RecordReader<IntValue> reader = new RecordReader<>(
+				getEnvironment().getInputGate(0),
+				IntValue.class,
+				getEnvironment().getTaskManagerInfo().getTmpDirectories());
+
+			IntValue i1 = reader.next();
+			IntValue i2 = reader.next();
+			IntValue i3 = reader.next();
+
+			if (i1.getValue() != 42 || i2.getValue() != 1337 || i3 != null) {
+				throw new Exception("Wrong data received.");
+			}
+		}
+	}
+
+	/**
+	 * This job runs in N slots with 2 * N senders and N receivers. Unless slot sharing is used,
+	 * it cannot complete.
+	 */
+	@Test
+	public void testSlotSharingForTwoInputsJob() throws Exception {
+		final int parallelism = 11;
+
+		final MiniClusterConfiguration cfg = new MiniClusterConfiguration.Builder()
+			.setNumTaskManagers(1)
+			.setNumSlotsPerTaskManager(parallelism)
+			.setConfiguration(getDefaultConfiguration())
+			.build();
+
+		try (final MiniCluster miniCluster = new MiniCluster(cfg)) {
+			miniCluster.start();
+
+			final JobVertex sender1 = new JobVertex("Sender1");
+			sender1.setInvokableClass(CountDownLatchedSender.class);
+			sender1.setParallelism(parallelism);
+
+			final JobVertex sender2 = new JobVertex("Sender2");
+			sender2.setInvokableClass(CountDownLatchedSender.class);
+			sender2.setParallelism(parallelism);
+
+			final JobVertex receiver = new JobVertex("Receiver");
+			receiver.setInvokableClass(CountDownLatchedAgnosticBinaryReceiver.class);
+			receiver.setParallelism(parallelism);
+
+			receiver.connectNewDataSetAsInput(sender1, DistributionPattern.POINTWISE,
+				ResultPartitionType.PIPELINED);
+			receiver.connectNewDataSetAsInput(sender2, DistributionPattern.ALL_TO_ALL,
+				ResultPartitionType.PIPELINED);
+
+			final CountDownLatch countDownLatch = new CountDownLatch(parallelism);
+			CountDownLatchedSender.setLatch(countDownLatch);
```

Comment:
this test must also not be run in parallel to `testSlotSharingForForwardJobWithCoLocationConstraint`

Reply:
Thanks! I add a `ReentrantLock` to guard settings to these `CountDownLatch`s.

Reply:
actually, the synchronization can be removed since multiple tests are not run in parallel.

Reply:
FYI: if the test had failed another test could've been blocked indefinitely since you aren't calling unlock in a finally block.

Reply:
hmmm thanks for you education and sorry for so inexperienced. will remove the synchronization

Reply:
does it make sense that we move unlock to finally block for possible parallel testing? or defer the synchronization until we really make it parallelized.